snuba/query/matchers.py

```diff
@@ -0,0 +1,317 @@
+
+
+@dataclass(frozen=True)
+class String(Pattern[str]):
+    """
+    Matches one specific string.
+    """
+
+    value: str
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class OptionalString(Pattern[Optional[str]]):
+    """
+    Matches one specific string (or None).
+    """
+
+    value: Optional[str]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class AnyOptionalString(Pattern[Optional[str]]):
+    """
+    Matches any string including the None value. This cannot be done with
+    Any(type) because that cannot match Union[str, None].
+    """
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        return MatchResult() if node is None or isinstance(node, str) else None
+
+
+@dataclass(frozen=True)
+class Or(Pattern[TMatchedType]):
+    """
+    Union of multiple patterns. Matches if at least one is a valid match
+    and returns the first valid one.
+    """
+
+    patterns: Sequence[Pattern[TMatchedType]]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        for p in self.patterns:
+            ret = p.match(node)
+            if ret:
+                return ret
+        return None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[ColumnExpr]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each, a Pattern can be
+    provided. If one of these Patterns is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[str]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, ColumnExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.table_name, node.table_name),
+        ):
+            if pattern is not None:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        # column_name does not have the same type as alias and table_name
+        # so this case cannot be handled in the loop above as it does not
+        # type check.
+        if self.column_name is not None:
+            partial_result = self.column_name.match(node.column_name)
+            if not partial_result:
+                return None
+            result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class Literal(Pattern[LiteralExpr]):
+    alias: Optional[Pattern[Optional[str]]] = None
+    value: Optional[Pattern[OptionalScalarType]] = None
```

Comment:
If `OptionalScalarType` is supposed to be symmetric with the type of `Literal.value` it'd probably be good to import it from there instead of redefining it in this file.